### DISCUSSION TITLE
Make shared state between Islands safe and enhance its use with View Transitions

### DISCUSSION THREAD

**Discussion by kevintakeda** [withastro/roadmap#756] [Proposal]
# Summary

This proposal aims to enable developers to safely share state across pages and components, as well as enhancing the transition between static components and those that require updates during View Transition navigation. The discussion about this started here: https://github.com/withastro/astro/issues/8781

Full demo of the problem: https://stackblitz.com/edit/github-bxb9uz-hyncuf?file=src%2Fpages%2Findex.astro

# Background & Motivation

Shared state brings a lot of convenience and power for Astro users, but as it is right now, it is dangerous to shoot oneself in the foot (E.g. Hydration mismatches) or make bad experiences (E.g. Saving the state of a full-page component with `client:only` causes scroll and screen flash issues when you return).

Firstly, the shared state is not safe. If the client modifies the shared state, any subsequent hydration may encounter mismatches. This occurs because the client's suggested DOM structure might differ from what the server initially sent. [Most UI Frameworks do not anticipate mismatches and will emit errors when they happen](https://github.com/withastro/astro/issues/8781#issuecomment-1786526962). From my tests Svelte and Lit seem to be the only frameworks that address this mismatch.

Currently, the approach to solve this problem is limited:
1. Use the same component on every page with `transition:persist`.
2. Use the directive `client:only`, which may lead to challenges reminiscent of "flash of white while using dark theme" issue and FOUC problem after View Transition and on first load.
3. Use only Svelte or Lit. However, flash and layout shift may occurr after View Transition.
4. Let the hydration happens, then use `useEffect` to replicate what happens in shared state to reflect on its scoped state. Seems to be the best solution for now, but also flash and layout shift may occurr after View Transition and require extra code.

Note: The shared state issue doesn't happen only with View Transition, [mismatches can happen at any point during the hydration](https://github.com/withastro/roadmap/discussions/756#discussioncomment-7452081). It just happens to occur more commonly when used with View Transition.

# Goals

- Allow Astro users to safely share state across pages and components.
- Prevent "Flash" or Layout shift due to the disparity between SSR and client state during View Transition navigations.
- Correctly restore the scroll after the component is fully rendered.

# Design

[I provided a detailed comment on the solution](https://github.com/withastro/astro/issues/8781#issuecomment-1787514581). In short, using the `render` function instead of the `hydrate` function on load and during the view transition.

A new directive can be introduced: `client:swap`, to mark a component that relies on a shared state to be rendered and swapped instead of hydrating. [This would ensure a partially hydrated page won't cause any mismatches.](https://github.com/withastro/roadmap/discussions/756#discussioncomment-7452081). This directive also instructs the component to render the component separately and seamlessly swap during navigation view transition swap stage.

In code:
```jsx
<ShoppingCart client:swap />
```

### Edge case
There is also the possibility to allow `swap` only after the component visible or the browser is idle, but I don't think is a good practice, `client:swap` implies the component already have a high priority. The component always needs to be rendered during view transitions to prevent flashes.

**Comment by kevintakeda** [comment]
I just realized, this problem doesn't happen only with View Transitions. It can happen at any point during the hydration. I just updated my proposal accordingly.
A minimal reproduction without using View Transitions: https://stackblitz.com/edit/github-bxb9uz-hhcjuw?file=src%2Fpages%2Findex.astro

**Comment by kevintakeda** [comment]
An alternative approach is to extend `addClientDirective` API to also allow rendering instead of hydrating.

```jsx
// client-myrender.js
export default async (load, opts, el) => {
  const trigger = await load()
  await trigger({ render: true })
}
```

```jsx
// In astro
<MyComponent client:myrender />
```

Since the client directive is a `Promise`, Astro can wait for the promise to finish after the swap inside `startViewTransition`.

**Comment by AustinFoss** [comment]
Was pointed in the direction of this proposal after submitting my own duplicate issue of this OP's , [withastro/astro#9893](https://github.com/withastro/astro/issues/9893)

Will keep posted on this and am eager to see how it progresses.

**Comment by matthewp** [comment]
Thanks for the details. Is this issue primarily about how certain frameworks handle hydration/rendering? Render retains state and hydration does not? If that's the case, maybe we can just always use render when "rerendering" a component?

I am a little confused though, since we added support for rerunning components on navigation with new props, and in testing the state was retaining there.

**Reply by kevintakeda** [reply]
> Is this issue primarily about how certain frameworks handle hydration/rendering?

Yes, most frameworks' hydration functions rely on the already rendered DOM, so they avoid recreating or updating it, even if the client shared state has changed.

> Render retains state and hydration does not?

Yes, but it's more accurate to say that the render reads the current client state and renders accordingly, and hydration assumes the state has not changed.

> If that's the case, maybe we can just always use render when "rerendering" a component?

Do you mean rendering always instead of hydrating?

That might work. I haven't benchmarked it, but there's some cost since you're always having to create DOM elements instead of reusing them. So it might increase the Time to Interactive (TTI) and Total Blocking Time (TBT) metric. So, it would be nice to allow users to choose between hydration and rendering, rather than forcing one or the other.

The timing to render things is also important. When rendering a component that might have a different size from the initial rendering can cause the scroll restoration to work incorrectly.

> I am a little confused though, since we added support for rerunning components on navigation with new props, and in testing the state was retaining there.

The shared state that I mentioned is not passed through props. Instead, it is just imported and used directly.

**Comment by jmurty** [comment]
We encountered this problem recently while experimenting with migrating our event ticketing website from Next.js to Astro.

Based on the very helpful suggestions by @kevintakeda [here](https://github.com/withastro/astro/issues/8781#issuecomment-1787514581) and [here](https://github.com/withastro/roadmap/discussions/756#discussioncomment-7461176) we found a workaround that solves the hydration errors for us, and could make our migration from Next.js to Astro feasible.

We add the following `client:render` [custom client directive](https://docs.astro.build/en/reference/integrations-reference/#addclientdirective-option) that forces/tricks Astro into **rendering** components in the browser instead of hydrating them, as if `client:only` was set, while the components are also rendered on the server (SSR) and our pages can be cached in the CDN as if `client:load` was set. It does this without needing any upstream changes in Astro itself.

```typescript
// astro-directive-client:render.ts
const clientRender: ClientDirective = async (load, opts, el) => {
    // Force `client:render` component to act like `client:only`: `hydrate()` function
    // will use `render` code path in the browser instead of `hydrate`.
    el.setAttribute("client", "only");

    // "Hydrate" as usual, will actually render not hydrate
    const hydrate = await load();
    await hydrate();
};
export default clientRender;

// Astro component shows tickets in a (P)React component, with availability info from client-side global state
<TicketListing client:render />
```

The key piece is the single line `el.setAttribute("client", "only");` that overwrites our custom `client:render` directive on the client side to look like `client:only` instead, before Asto runs its hydrate/render process.

This is definitely a bit of a hack, but it works for us with the latest Astro 5 and 6 versions and could unblock our migration of our website from Next.js to Astro.

Can anyone foresee any glaring problems this `client:render` approach might cause?

And would there be interest in making this SSR + always-render-on-client approach an officially supported option in Astro? 
I am not yet very familiar with the codebase, but I could probably sketch out an initial PR for review if there is interest?

---

For context, our site needs to maintain a lot of client-side global state to track things like ticket availability (sold out, almost sold out) plus the customer's changing cart contents, and to display this info in different ways across multiple pages and components. We want to do this by fetching frequently changing data from APIs without the need for server-side re-rendering of pages to reflect constantly changing ticket availability. And we want server pages to be pre-rendered for SEO, but to be cached medium-term and served from a CDN, not dynamically rendered, for a faster website experience and reduced cost for running server-side functions to re-render pages.

We do all of this in Next.js already with client global state stored in React Query's client cache with React components. And this same approach worked well in Astro after migration – with very similar code – but only for initial page loads + hydration. Navigating to pages after the initial page load via `<ClientRouter />` caused the kinds of hydration failures outlined in this and related issues.
